lib/hello/catalog.ex

```diff
@@ -0,0 +1,229 @@
+defmodule Hello.Catalog do
+  @moduledoc """
+  The Catalog context.
+  """
+
+  import Ecto.Query, warn: false
+  alias Hello.Repo
+
+  alias Hello.Catalog.Product
+  alias Hello.Catalog.Category
+
+  @doc """
+  Returns the list of products.
+
+  ## Examples
+
+      iex> list_products()
+      [%Product{}, ...]
+
+  """
+  def list_products do
+    Repo.all(Product)
+  end
+
+  @doc """
+  Gets a single product.
+
+  Raises `Ecto.NoResultsError` if the Product does not exist.
+
+  ## Examples
+
+      iex> get_product!(123)
+      %Product{}
+
+      iex> get_product!(456)
+      ** (Ecto.NoResultsError)
+
+  """
+  #def get_product!(id), do: Repo.get!(Product, id)
+  # We load the categories for a specific product
+  def get_product!(id) do
+    Product
+    |> Repo.get(id)
+    |> Repo.preload(:categories)
+  end
+
+  @doc """
+  Creates a product.
+
+  ## Examples
+
+      iex> create_product(%{field: value})
+      {:ok, %Product{}}
+
+      iex> create_product(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def create_product(attrs \\ %{}) do
+    %Product{}
+    #|> Product.changeset(attrs)
+    |> change_product(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Updates a product.
+
+  ## Examples
+
+      iex> update_product(product, %{field: new_value})
+      {:ok, %Product{}}
+
+      iex> update_product(product, %{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def update_product(%Product{} = product, attrs) do
+    product
+    #|> Product.changeset(attrs)
+    |> change_product(attrs)
+    |> Repo.update()
+  end
+
+  @doc """
+  Deletes a product.
+
+  ## Examples
+
+      iex> delete_product(product)
+      {:ok, %Product{}}
+
+      iex> delete_product(product)
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def delete_product(%Product{} = product) do
+    Repo.delete(product)
+  end
+
+  @doc """
+  Returns an `%Ecto.Changeset{}` for tracking product changes.
+
+  ## Examples
+
+      iex> change_product(product)
+      %Ecto.Changeset{data: %Product{}}
+
+  """
+  def change_product(%Product{} = product, attrs \\ %{}) do
+    #Product.changeset(product, attrs)
+    categories = list_categories_by_id(attrs["category_ids"])
+
+    product
+    |> Repo.preload(:categories)
+    |> Product.changeset(attrs)
+    |> Ecto.Changeset.put_assoc(:categories, categories)
```

Comment:
@keliumJU what is a changeset?

Reply:
Changeset is a module that mainly helps us to validate the properties of the model, by means of a structure that describes what errors our data presents (name, age, gender...), so we make sure to carry out an action in a clean way.